src/phoebus_utils/programming_utils.hpp

```diff
@@ -0,0 +1,49 @@
+// © 2022. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract
+// 89233218CNA000001 for Los Alamos National Laboratory (LANL), which
+// is operated by Triad National Security, LLC for the U.S.
+// Department of Energy/National Nuclear Security Administration. All
+// rights in the program are reserved by Triad National Security, LLC,
+// and the U.S. Department of Energy/National Nuclear Security
+// Administration. The Government is granted for itself and others
+// acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works,
+// distribute copies to the public, perform publicly and display
+// publicly, and to permit others to do so.
+
+#ifndef PHOEBUS_UTILS_PROGRAMMING_UTILS_HPP_
+#define PHOEBUS_UTILS_PROGRAMMING_UTILS_HPP_
+
+#include <algorithm>
+#include <vector>
+
+namespace programming {
+
+template <class SPECIAL, template <class...> class TEMPL>
+struct is_specialization_of : public std::false_type {};
+
+template <template <class...> class TEMPL, class... TPARAMS>
+struct is_specialization_of<TEMPL<TPARAMS...>, TEMPL> : public std::true_type {};
+
+template <typename T>
+bool soft_equiv(const T x, const T y, const T tolerance = static_cast<T>(1.e-10f)) {
+  bool equivalent = false;
+  if (2. * fabs(y - x) < tolerance * (std::fabs(x) + std::fabs(y))) {
+    equivalent = true;
+  }
+
+  return equivalent;
+}
+
+template <typename T>
+bool ContainedInVector(T &item, std::vector<T> &container) {
```

Comment:
What is the use case for this? If you want to make calls like this regularly, maybe a `std::vector` is the wrong container and a set or map is better?

Reply:
Right now I use this for checking string arguments against a vector of acceptable values (e.g. radiation method):
```c++
  std::vector<std::string> known_methods = {"cooling_function", "moment_m1",                         
                                            "moment_eddington", "monte_carlo", "mocmc"};             
  if (!programming::ContainedInVector(method, known_methods)) {                                      
    std::stringstream msg;                                                                           
    msg << "Radiation method \"" << method << "\" not recognized!";                                  
    PARTHENON_FAIL(msg);                                                                             
  }
  ```
  Maybe this could automatically work with a set or map just by doing something like
  ```c++
template <template <typename> class V, typename T>                                                   
bool ContainedInVector(T &item, V<T> &container) {
```
-- I tried the above but it didn't work so I'm moving on for now

Reply:
OOOH I see. Okay this is actually pretty useful then.

Your example should be this:
```C++
template <template <typename> class V, typename T, typename... Args>                                                   
bool ContainedInVector(T &item, V<T, Args...> &container) {
```
because the STL containers often have a bunch of hidden default template parameters.

Doesn't matter though. I only argued for a different type because I didn't realize what you were doing here.

Reply:
Yeah something about the `std::find` syntax makes me want to avoid it. Between that and `std::vector<bool>` being broken, I think it is enough justification for us to have our own `pvector` class in here that mostly just wraps `std::vector` but adds, like, a `.contains(T x)` method 🤷

Reply:
I think we have `std::vector` all over the place in `parthenon`, so this may not be practical, but `std::set` is ordered, unique, and has a `contains` method.

Reply:
I just looked -- isn't `std::set::contains` a C++20 feature? https://en.cppreference.com/w/cpp/container/set/contains

Reply:
Yes but you can just use `std::set::count` and treat it as a boolean.